src/spyglass/utils/h5py_helper_fn.py

```diff
@@ -0,0 +1,247 @@
+from __future__ import annotations
+
+import uuid
+from pathlib import Path
+from typing import List, Optional
+
+import h5py
+import numpy as np
+
+
+def is_float16_dtype(
+    dt: np.dtype, *, include_subarray_base: bool = False
+) -> bool:
+    """Return True if dtype is a plain float16 (any endianness).
+
+    Parameters
+    ----------
+    dt : numpy.dtype
+        Dtype to test.
+    include_subarray_base : bool, optional
+        If True, treat subarray dtypes with base float16 as float16.
+        If False (default), exclude subarray dtypes.
+
+    Returns
+    -------
+    bool
+        True if dtype is float16 under the chosen policy; otherwise False.
+
+    Notes
+    -----
+    Excludes:
+    - Object/region references
+    - Variable-length (vlen) types
+    - Compound dtypes
+    - Subarray dtypes unless `include_subarray_base=True`
+    """
+    # Exclude references and vlen types
+    if h5py.check_dtype(ref=dt) is not None:
+        return False
+    if h5py.check_dtype(vlen=dt) is not None:
+        return False
+
+    # Exclude compound
+    if dt.names is not None:
+        return False
+
+    # Handle subarray dtypes (e.g., (M, N)<f2)
+    if dt.subdtype is not None:
+        if not include_subarray_base:
+            return False
+        base, _ = dt.subdtype
+        dt = np.dtype(base)
+
+    # Match binary16 (IEEE 754 half, <f2/>f2/=f2)
+    return dt.kind == "f" and dt.itemsize == 2
+
+
+def find_float16_datasets(
+    file_path: str,
+    *,
+    include_dimension_scales: bool = False,
+    include_subarray_base: bool = False,
+) -> List[str]:
+    """List absolute HDF5 paths of datasets stored as float16.
+
+    Parameters
+    ----------
+    file_path : str
+        Path to the HDF5/NWB file.
+    include_dimension_scales : bool, optional
+        If False (default), skip datasets that are HDF5 dimension scales
+        (CLASS='DIMENSION_SCALE').
+    include_subarray_base : bool, optional
+        If True, include subarray dtypes whose base is float16.
+
+    Returns
+    -------
+    list of str
+        Absolute dataset paths with dtype float16 under the chosen policy.
+    """
+    hits: List[str] = []
+
+    def _visit(name: str, obj) -> None:
+        if not isinstance(obj, h5py.Dataset):
```

Comment:
Some of the current h5 code is doing this 'if is type, visit' logic, though embedded in the Comparison class